### PR TITLE
[ValueTracking] Compute knownbits from `(icmp upred (add/sub nuw X, Y), C)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for computing knownbits from `(icmp upred (add/sub nuw X, Y), C)`; NFC**
- **[ValueTracking] Compute knownbits from `(icmp upred (add/sub nuw X, Y), C)`**

    `(icmp ule/ult (add nuw X, Y), C)` implies both `(icmp ule/ult X, C)` and
    `(icmp ule/ult Y, C)`. We can use this to deduce leading zeros in `X`/`Y`.
    
    `(icmp uge/ugt (sub nuw X, Y), C)` implies `(icmp uge/uge X, C)` . We
    can use this to deduce leading ones in `X`.
    
    Proofs: https://alive2.llvm.org/ce/z/sc5k22

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -9283,11 +9289,21 @@ void llvm::findValuesAffectedByCondition(
...
-          AddAffected(X);
+        if (match(B, m_ConstantInt())) {
+          // Handle (A + C1) u< C2, which is the canonical form of
+          // A > C3 && A < C4.
+          if (match(A, m_AddLike(m_Value(X), m_ConstantInt())))
+            AddAffected(X);
+
+          Value *Y;
+          // X & Y u> C -> X >u C && Y >u C
+          // X | Y u< C -> X u< C && Y u< C
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:9300]: Please update the comments.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Comment by dtcxzyw** [comment]
Can you have a look at the regression https://github.com/dtcxzyw/llvm-opt-benchmark/pull/465#discussion_r1546021756?

**Comment by goldsteinn** [comment]
> Can you have a look at the regression [dtcxzyw/llvm-opt-benchmark#465 (comment)](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/465#discussion_r1546021756)?

Okay, reduced test:
```
define void @foo(i32 %x, i32 %y) {
  %z = add nuw i32 %x, %y
  %cmp0 = icmp ult i32 %z, 255
  br i1 %cmp0, label %L0, label %L2

L0:
  %sub_x = sub i32 255, %x
  br label %L1

L1:
  %phi = phi i32 [ 0, %L0 ], [ 1, %L1 ]
  %cmp1 = icmp ult i32 %phi, %sub_x
  br i1 %cmp1, label %L1, label %L2

L2:
  ret void
}
```

The issue occurs due to the InstCombine fold at:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp#L2280-L2288

We produce:

**Bad:**
```
define void @foo(i32 %x, i32 %y) local_unnamed_addr {
  %z = add nuw i32 %x, %y
  %cmp0 = icmp ult i32 %z, 255
  br i1 %cmp0, label %L0, label %L2

L0:                                               ; preds = %0
  %sub_x = xor i32 %x, 255
  br label %L1

L1:                                               ; preds = %L1, %L0
  %phi = phi i32 [ 0, %L0 ], [ 1, %L1 ]
  %cmp1 = icmp ult i32 %phi, %sub_x
  br i1 %cmp1, label %L1, label %L2

L2:                                               ; preds = %L1, %0
  ret void
}
```

vs
**Good:**
```
define void @foo(i32 %x, i32 %y) local_unnamed_addr {
  %z = add nuw i32 %x, %y
  %cmp0 = icmp ult i32 %z, 255
  br i1 %cmp0, label %L0, label %L2

L0:                                               ; preds = %0
  %sub_x = sub i32 255, %x
  br label %L1

L1:                                               ; preds = %L1, %L0
  %phi = phi i32 [ 0, %L0 ], [ 1, %L1 ]
  %cmp1 = icmp ult i32 %phi, %sub_x
  br i1 %cmp1, label %L1, label %L2

L2:                                               ; preds = %L1, %0
  ret void
}
```

The issue then arises because we fold `(icmp eq/ne (sub C0, x), C1)` -> `(icmp eq/ne x, C2)` even if the sub has multi-use but we don't do the same for `xor`:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L3463-L3474

Ill add a patch for the `xor` case.

**Comment by goldsteinn** [comment]
My preference would be to not have this blocked by the missing multi-use xor fold. It seems pretty unrelated that we end up with that pattern w/ this patch but not w/ (in that one case) and handling multi-use xor seems independently valuable.

**Comment by goldsteinn** [comment]
@dtcxzyw i update with support for `sub nuw` as well.

**Comment by goldsteinn** [comment]
rebased

**Comment by nikic** [comment]
Needs another rebase.

> @dtcxzyw i update with support for `sub nuw` as well.

Why?

**Comment by goldsteinn** [comment]
> Needs another rebase.
> 
> > @dtcxzyw i update with support for `sub nuw` as well.
> 
> Why?

I guess the `add nuw` mirrors the `or` case and the `sub nuw` mirrors the `and` case. I had meant to leave it as a folllowup todo but it seemed to fit well enough.

**Comment by goldsteinn** [comment]
ping

**Comment by goldsteinn** [comment]
ping

File: llvm/test/Transforms/InstCombine/known-bits.ll [hunk 2]
```diff
@@ -1374,5 +1374,226 @@ define i8 @nonzero_reduce_xor_vscale_odd(<vscale x 3 x i8> %xx) {
...
+  %cmp = icmp sge i8 %n_sub, 32
+  br i1 %cmp, label %if.then, label %if.else
+
+if.then:
+  %r = and i8 %n, 32
+  ret i8 %r
+
+if.else:
+  ret i8 %other
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/known-bits.ll:1596]: Negative tests without nuw?

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/InstCombine/known-bits.ll (hunk 2 above)
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/known-bits.ll]: done + rebased/

**Pull request closed without merging**